File: algo/binary_search/binary_search.js

const binarySearchRecursive = (arr, start, finish, target) => {

  if (finish >= start){
     
    let mid = start + (finish - start) / 2
    let middle_element = arr[mid]

    if(middle_element === target){

      return mid

    }else if(middle_element > target){

      return binarySearch(arr, start, mid - 1, target)

    }else{
      
      return binarySearch(arr, mid + 1, finish, target)

    }
  }else{
    return -1
  }
  
}

const binarySearchIterative = (array, target) => {

  let i = 0;
  let j = array.length-1;

  while(j >= i){
    let mid = i + (j - i)/2
    
    if(array[mid] === target){

      return mid

    }else if(array[mid] > target){

      j = mid - 1

    }else if(array[mid] < target){

      i = mid + 1

    }
  }

  return -1

}

File: algo/binary_search/binary_search.rb
def binary_search_recursive(array, start_idx, end_idx, target)

  if end_idx >= start_idx

    mid = start_idx + (end_idx - start_idx)/2

    if array[mid] == target

      return mid

    elsif array[mid] > target

      return binary_search_recursive(array, start_idx, mid - 1, target)

    elsif array[mid] < target

      return binary_search_recursive(array, mid + 1, end_idx, target)

    end
  else
    return nil
  end

end

def binary_search_iterative(array, start_idx, end_idx, target)
  
  while end_idx >= start_idx
    mid = start_idx + (end_idx - start_idx)/2

    if array[mid] == target
      return  mid
    elsif array[mid] < target
      start_idx = mid + 1
    else
      end_idx = mid - 1
    end
  end

  puts nil
  
end


arr = [ 2, 3, 4, 10, 40 ]
x = 10

puts binary_search_iterative(arr, 0, arr.length - 1, x)